numpy/core/tests/test_multiarray.py

```diff
@@ -1357,6 +1357,34 @@ def test_multiindex_titles(self):
         assert_raises(ValueError, lambda : a[['b','b']])  # field exists, but repeated
         a[['b','c']]  # no exception
 
+    def test_structured_cast_promotion_fieldorder(self):
+        # gh-15494
+        # dtypes with different field names are not promotable
+        A = ("a", "<i8")
+        B = ("b", "<i8")
+        ab = np.array([(1, 2)], dtype=[A, B])
+        ba = np.array([(1, 2)], dtype=[B, A])
+        assert_raises(TypeError, np.concatenate, [ab, ba])
```

Comment:
What's the behavior of the following?
```suggestion
        assert_raises(TypeError, np.concatenate, [ab, ba])
        assert_raises(TypeError, np.concatenate, [ab[['a', 'b']], ba[['a', 'b']]])
```

Reply:
Ah, I guess that's covered below